### ISSUE TITLE
support `__slots__`

### ISSUE THREAD

**Issue by mtshiba** [astral-sh/ty#1268]
`__slots__` is already handled specially in ty, but should be more fully supported in line with the Python specification.
`__slots__` may not be a very commonly used feature, but several major libraries use it (e.g. [attrs](https://github.com/python-attrs/attrs/blob/50334f3fe759013f52cc533161e9b474e98bc460/src/attr/_make.py#L2436)), and supporting it could improve inference results and error reporting.

ref: https://docs.python.org/3/reference/datamodel.html#slots

Here are the features that should be implemented:

- [ ] Assume that the attributes defined in `__slots__` are not unresolved
- [ ] Attempting to read/write/declare an attribute that is not defined in `__slots__` will be reported as an error
- [ ] Defining a class variable to set a default value for a name that is defined in `__slots__` will be reported as an error
- [ ] Remove `__dict__` and `__weakref__` from classes that define `__slots__`
- [ ] An error will be raised if nonempty `__slots__` are defined for a class derived from a "variable-length" built-in type such as int, bytes, and tuple

```python
class C:
    __slots__ = ("foo",)

    def __init__(self, foo, bar):
        self.foo = foo
        self.bar = bar  # error

    baz: int  # error (pyright doesn't report this as an error)
    foo: int = 1  # error

class D:
    __slots__ = ("bar",)

d = D()
reveal_type(d.bar)  # revealed: Any or Unknown
reveal_type(d.foo)  # error
d.bar = 1  # ok
# Classes with `__slots__` defined will not have a `__dict__` by default.
d.foo = 1  # error

class E:
    __slots__ = ("foo", "__dict__")

    def __init__(self, foo, bar):
        self.foo = foo
        self.bar = bar  # ok

# `F` has `__dict__` and `__weakref__`.
class F: ...

# Even if a class defines `__slots__`, if the superclass has `__dict__` or `__weakref__`,
# those attributes will be inherited.
class G(F):
    __slots__ = ("foo",)

    def __init__(self, foo, bar):
        self.foo = foo
        self.bar = bar  # ok

# error: nonempty __slots__ not supported for subtype of 'int'
class H(int):
    __slots__ = ("foo",)
```

related: #111

**Comment by pmhahn** [comment]
```python
o = object()
if hasattr(o, "__slots__"):
  s = o.__slots__
  reveal_type(s)
```

- ty: `object` ❓ 
- mypy: `Any`
- pyrefly: `Unknown`
- pyright: `Unknown` + `error: Cannot access attribute "__slots__" for class "object"`

https://docs.python.org/3/reference/datamodel.html#object.__slots__
> This class variable can be assigned a string, iterable, or sequence of strings

`object` does not look right.

**Comment by carljm** [comment]
@pmhahn This isn't specific to `__slots__`, it's just about `hasattr` narrowing behavior. `__slots__` isn't a typed attribute of Python objects in typeshed. You can replace `__slots__` in this example with `any_random_attr` and see exactly the same list of type checker behaviors.

So implicitly there are two questions here:
1. Should ty be lenient by default and assume `Unknown` instead of `object` as the type of a `hasattr` provided attribute? (This should be a separate issue, it's off topic here.)
2. Should ty (and other type checkers) special case the attribute name `__slots__` and give it a more precise type, if it's present? That seems reasonable, and I guess it's higher value for ty than other type checkers, since other type checkers will just silently let you do whatever with `Unknown/Any`.

**Comment by pmhahn** [comment]
@carljm I'm currently working on https://github.com/eliben/pyelftools/pull/611 and found code, which `mypy`, `pyright`, `pyrefly` are happy with, but `ty` complained about — actually `ty` is right and there is a bug in the code, but for other reasons. So 👍 for `ty`!

I would expect something like
`__slots__: str | Sequence[str] | Mapping[str, str] | Iterator[str]` or just `str | Iterable[str]` as Python also accepts `set[str]`.

But `object` was/is unexpected (by me) and I would appreciate a more precise type.

**Comment by jonathandung** [comment]
After support for this is added, would it be beneficial to add an opt-in rule that flags declarations of `__slots__` in classes, at least one base class of which does not declare `__slots__`, since the benefit of avoiding the instance dictionary would be gone?
